src/psygnal/_group.py

```diff
@@ -237,17 +190,170 @@ def disconnect(self, slot: Callable | None = None, missing_ok: bool = True) -> N
+
+    This class is used in both the EventedModels and the evented dataclass patterns.
+    See also: `psygnal.SignalGroupDescriptor`, which provides convenient and explicit
+    way to create a SignalGroup on a dataclass-like class.
+
+    Parameters
+    ----------
+    instance : Any, optional
+        An object to which this `SignalGroup` is bound, by default None
+
+    Attributes
+    ----------
+    all : SignalRelay
+        A special SignalRelay instance that can be used to connect to all signals in
+        this group.  The name of this attribute can be overridden by the user by
+        creating a new name for the SignalRelay annotation on a subclass of SignalGroup
+        e.g. `my_name: SignalRelay`
+
+    Examples
+    --------
+    ```python
+    from psygnal import Signal, SignalGroup
+
+    class MySignals(SignalGroup):
+        sig1 = Signal()
+        sig2 = Signal()
+
+    group = MySignals()
+    group.all.connect(print) # connect to all signals in the group
+
+    list(group)                  # ['sig1', 'sig2']
+    len(group)                   # 2
+    group.sig1 is group['sig1']  # True
+    ```
+    """
+
+    _psygnal_signals: ClassVar[Mapping[str, Signal]]
+    _psygnal_instances: dict[str, SignalInstance]
+    _psygnal_uniform: ClassVar[bool] = False
+
+    # see comment in __init__.  This type annotation can be overriden by subclass
+    # to change the public name of the SignalRelay attribute
+    all: SignalRelay
+
+    def __init__(self, instance: Any = None) -> None:
+        cls = type(self)
+        if not hasattr(cls, "_psygnal_signals"):  # pragma: no cover
+            raise TypeError(
+                "Cannot instantiate SignalGroup directly.  Use a subclass instead."
+            )
+        self._psygnal_instances: dict[str, SignalInstance] = {
+            name: signal.__get__(self, cls)
+            for name, signal in cls._psygnal_signals.items()
+        }
+        self._psygnal_relay = SignalRelay(self._psygnal_instances, instance)
+
+        # determine the public name of the signal relay.
+        # by default, this is "all", but it can be overridden by the user by creating
+        # a new name for the SignalRelay annotation on a subclass of SignalGroup
+        # e.g. `my_name: SignalRelay`
+        self._psygnal_relay_name = "all"
+        for base in cls.__mro__:
+            for key, val in getattr(base, "__annotations__", {}).items():
+                if val is SignalRelay:
+                    self._psygnal_relay_name = key
+                    break
+        setattr(self, self._psygnal_relay_name, self._psygnal_relay)
+
+    def __init_subclass__(cls, strict: bool = False) -> None:
+        """Collects all Signal instances on the class under `cls._psygnal_signals`."""
+        cls._psygnal_signals = {
+            k: val
+            for k, val in getattr(cls, "__dict__", {}).items()
+            if isinstance(val, Signal)
+        }
+
```

Comment:
Add a mechanism to catch potential name conflicts early on.
The reserved attributes can be the one starting with `_psygnal_` or a reduced list (in this case we may have conflicts if we add private attributes to SignalGroup later.

```python
        conflict_keys = set(k for k in cls._psygnal_signals if k.startswith("_psygnal_"))
        if len(conflict_keys) > 0:
            warnings.warn(
                "Signal names are conflicting with SignalGroup private attribute, "
                f"skip signals: {conflict_keys}",
                stacklevel=2,
            )
            for key in conflict_keys:
                del cls._psygnal_signals[key]
```

Reply:
i like it, thanks

Reply:
I think you forgot to add this :)

Reply:
added thanks